### PR TITLE
Revamp encoding-decoding with KotlinXSerializationMapper

### PR DESCRIPTION
**Pull request by fscarponi**
The previous implementation of Document Encoding was missing Polymorphism support for maps and lists, on top of that there were many corner cases not handled.
Given the similarity between JSON and Documents, it makes sense to leverage on KotlinxJsonSerialization implementation and convert JsonObjects to Documents.
This allows having the same degree of freedom as JsonSerialization

### CODE REVIEW

File: potassium-nitrite/src/main/kotlin/org/dizitart/kno2/serialization/DocumentFormat.kt [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    is JsonObject -> toDocument()
+    is JsonArray -> toDocument()
+    is JsonNull -> null
+    is JsonPrimitive -> toDocument()
+}
+
+private fun JsonObject.toDocument(): Document {
+    val document = Document.createDocument()
+    for ((key, value) in entries) {
+        document.put(key, value.toDocument(), true)
```
Comment (anidotnet) [inline potassium-nitrite/src/main/kotlin/org/dizitart/kno2/serialization/DocumentFormat.kt:160]: This is very much your use case specific change. If some other user choose to use this kotlinx.serialization they won't be able to use/query nested documents. Please see my comments [here](https://github.com/nitrite/nitrite-java/pull/968#discussion_r1603752011).

File: potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt [hunk 2]
```diff
@@ -43,7 +45,7 @@ class KotlinXSerializationMapperTest {
         val someMap: Map<String, String>,
         val someSerializableObjectMap: Map<InnerObject, SomePolymorphType>,
         val innerObject: InnerObject,
-        val id: String,
+        @SerialName("_id") val id: String? = null,
```
Comment (anidotnet) [inline potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt:51]: I would recommend not to name any user defined field as "_id". It is reserved to store the `NitriteId` in a document. 

If you want to get hold of the `NitriteId`, just use the datatype as `NitriteId` and mark the field with `@Id` annotation, any field name would be ok. 

If not, keep the field name anything other than "_id" in the document.

File: potassium-nitrite/src/main/kotlin/org/dizitart/kno2/serialization/DocumentFormat.kt (hunk 1 above)
Comment (fscarponi) [reply potassium-nitrite/src/main/kotlin/org/dizitart/kno2/serialization/DocumentFormat.kt]: let's use [this comment](https://github.com/nitrite/nitrite-java/pull/968#discussion_r1603752011) to follow up the topic!

File: potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt (hunk 2 above)
Comment (fscarponi) [reply potassium-nitrite/src/test/kotlin/org/dizitart/kno2/KotlinXSerializationMapperTest.kt]: Got it!

**Comment by anidotnet** [comment]
Closing the PR for lack of activity. Feel free to reopen for further changes.

**Pull request closed without merging**